sklearn/ensemble/iforest.py

```diff
@@ -0,0 +1,245 @@
+        If int, random_state is the seed used by the random number generator;
+        If RandomState instance, random_state is the random number generator;
+        If None, the random number generator is the RandomState instance used
+        by `np.random`.
+
+    verbose : int, optional (default=0)
+        Controls the verbosity of the tree building process.
+
+
+    Attributes
+    ----------
+    estimators_ : list of DecisionTreeClassifier
+        The collection of fitted sub-estimators.
+
+    estimators_samples_ : list of arrays
+        The subset of drawn samples (i.e., the in-bag samples) for each base
+        estimator.
+
+    max_samples_ : integer
+        The actual number of samples
+
+    References
+    ----------
+    .. [1] Liu, Fei Tony, Ting, Kai Ming and Zhou, Zhi-Hua. "Isolation forest."
+           Data Mining, 2008. ICDM'08. Eighth IEEE International Conference on.
+    .. [2] Liu, Fei Tony, Ting, Kai Ming and Zhou, Zhi-Hua. "Isolation-based
+           anomaly detection." ACM Transactions on Knowledge Discovery from
+           Data (TKDD) 6.1 (2012): 3.
+    """
+
+    def __init__(self,
+                 n_estimators=100,
+                 max_samples=256,
+                 max_features=1.,
+                 bootstrap=True,
+                 n_jobs=1,
+                 random_state=None,
+                 verbose=0):
+        super(IsolationForest, self).__init__(
+            base_estimator=ExtraTreeRegressor(
+                max_depth=int(np.ceil(np.log2(max(max_samples, 2)))),
+                max_features=1,
+                splitter='random',
+                random_state=random_state),
+            # here above max_features has no links with self.max_features
+            bootstrap=bootstrap,
+            bootstrap_features=False,
+            n_estimators=n_estimators,
+            max_samples=max_samples,
+            max_features=max_features,
+            n_jobs=n_jobs,
+            random_state=random_state,
+            verbose=verbose)
+
+    def _set_oob_score(self, X, y):
+        raise NotImplementedError("OOB score not supported by iforest")
+
+    def fit(self, X, y=None, sample_weight=None):
+        """Fit estimator.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix, shape (n_samples, n_features)
+            The input samples. Use ``dtype=np.float32`` for maximum
+            efficiency. Sparse matrices are also supported, use sparse
+            ``csc_matrix`` for maximum efficieny.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        # ensure_2d=False because there are actually unit test checking we fail
+        # for 1d.
+        X = check_array(X, accept_sparse=['csc'], ensure_2d=False)
+        if issparse(X):
+            # Pre-sort indices to avoid that each individual tree of the
+            # ensemble sorts the indices.
+            X.sort_indices()
```

Comment:
We should make a _fit_validate in the tree code.
Though this may be done elsewhere.